optaplanner-wb-screens/optaplanner-wb-solver-editor/optaplanner-wb-solver-editor-client/src/main/java/org/optaplanner/workbench/screens/solver/client/editor/TerminationTreeItemContent.java

```diff
@@ -15,84 +15,408 @@
  */
 package org.optaplanner.workbench.screens.solver.client.editor;
 
-import com.google.gwt.dom.client.Style;
-import com.google.gwt.user.client.ui.Composite;
-import com.google.gwt.user.client.ui.HorizontalPanel;
-import com.google.gwt.user.client.ui.Label;
+import java.util.HashMap;
+import java.util.Map;
+import javax.enterprise.context.Dependent;
+import javax.inject.Inject;
+
+import com.google.gwt.core.client.GWT;
+import com.google.gwt.user.client.ui.IsWidget;
 import com.google.gwt.user.client.ui.TreeItem;
-import org.gwtbootstrap3.client.ui.Button;
-import org.gwtbootstrap3.client.ui.constants.ButtonType;
-import org.gwtbootstrap3.client.ui.constants.IconType;
+import com.google.gwt.user.client.ui.Widget;
 import org.optaplanner.workbench.screens.solver.client.resources.i18n.SolverEditorConstants;
+import org.optaplanner.workbench.screens.solver.model.TerminationCompositionStyleModel;
 import org.optaplanner.workbench.screens.solver.model.TerminationConfigModel;
 import org.optaplanner.workbench.screens.solver.model.TerminationConfigOption;
 
-public class TerminationTreeItemContent extends Composite {
+import static org.optaplanner.workbench.screens.solver.model.TerminationConfigOption.*;
 
-    private Label terminationTypeLabel = new Label();
-    private Label terminationValueLabel = new Label();
-    private Button addTerminationButton = new Button();
-    private Button removeTerminationButton = new Button();
+@Dependent
+public class TerminationTreeItemContent implements IsWidget {
```

Comment:
I agree with @manstis. The goal is to use as a few widget types as possible. I think that part we've communicated. The best practice is shaping up exactly how Michael described. @ederign let's chat on Fri. or early next week and provide an example somewhere to share with everyone.